cpp/src/neighbors/detail/cagra/cagra_merge.cuh

```diff
@@ -0,0 +1,130 @@
+index<T, IdxT> merge(raft::resources const& handle,
+                     const cagra::merge_params& params,
+                     std::vector<cuvs::neighbors::cagra::index<T, IdxT>*>& indices)
+{
+  std::size_t dim              = 0;
+  std::size_t new_dataset_size = 0;
+  int64_t stride               = -1;
+
+  for (auto index : indices) {
+    using ds_idx_type = decltype(index->data().n_rows());
+    if (auto* strided_dset = dynamic_cast<const strided_dataset<T, ds_idx_type>*>(&index->data());
+        strided_dset != nullptr) {
+      if (dim == 0) {
+        dim    = index->dim();
+        stride = strided_dset->stride();
+      } else {
+        RAFT_EXPECTS(dim == index->dim(), "Dimension of datasets in indices must be equal.");
+      }
+      new_dataset_size += index->size();
+    } else if (dynamic_cast<const cuvs::neighbors::empty_dataset<int64_t>*>(&index->data()) !=
+               nullptr) {
+      RAFT_FAIL(
+        "cagra::merge only supports an index to which the dataset is attached. Please check if the "
+        "index was built with index_param.attach_dataset_on_build = true, or if a dataset was "
+        "attached after the build.");
+    } else {
+      RAFT_FAIL("cagra::merge only supports an uncompressed dataset index");
+    }
+  }
+
+  auto host_updated_dataset = raft::make_host_matrix<T, std::int64_t>(new_dataset_size, dim);
+  memset(host_updated_dataset.data_handle(), 0, sizeof(T) * host_updated_dataset.size());
+
+  IdxT offset = 0;
+
+  for (auto index : indices) {
+    using ds_idx_type  = decltype(index->data().n_rows());
+    auto* strided_dset = dynamic_cast<const strided_dataset<T, ds_idx_type>*>(&index->data());
+
+    RAFT_CUDA_TRY(cudaMemcpy2DAsync(host_updated_dataset.data_handle() + offset * dim,
+                                    sizeof(T) * dim,
+                                    strided_dset->view().data_handle(),
+                                    sizeof(T) * stride,
+                                    sizeof(T) * dim,
+                                    strided_dset->n_rows(),
+                                    cudaMemcpyDefault,
+                                    raft::resource::get_cuda_stream(handle)));
+
+    offset += IdxT(index->data().n_rows());
+  }
+  // Allocate the new dataset on device
+  auto device_updated_dataset =
+    raft::make_device_matrix<T, std::int64_t>(handle, new_dataset_size, dim);
+  auto device_updated_dataset_view = raft::make_device_matrix_view<T, std::int64_t>(
+    device_updated_dataset.data_handle(), new_dataset_size, dim);
+
+  // Copy updated dataset on host memory to device memory
+  raft::copy(device_updated_dataset.data_handle(),
+             host_updated_dataset.data_handle(),
+             new_dataset_size * dim,
+             raft::resource::get_cuda_stream(handle));
+
+  auto merged_index =
+    cagra::build(handle, params, raft::make_const_mdspan(device_updated_dataset_view));
+
+  if (static_cast<std::size_t>(stride) == dim) {
```

Comment:
Hi @cjnolet @achirkin , I know these codes are odd, but without them, datasets will be changed after calling `cagra::detail::search_main_core,` which will cause the test failure.  I do not know how the `dataset format, matrix ownership, cagra::search` interact behind it. Could you have comments here? Many thanks!

Reply:
To me, it looks like no one owns the `host_updated_dataset` or `device_updated_dataset` beyond the scope of this function, so the data gets destroyed unless the owning `update_dataset` is called under the `if` branch here.
Hence, I think, you should call `update_dataset` unconditionally here.

Reply:
> To me, it looks like no one owns the `host_updated_dataset` or `device_updated_dataset` beyond the scope of this function, so the data gets destroyed unless the owning `update_dataset` is called under the `if` branch here. Hence, I think, you should call `update_dataset` unconditionally here.

Thank you, very helpful!

Reply:
Hi @achirkin , I got the reason, the `update_dataset` is called in `cagra::build` potentially, just when `stride==dim` , a non-owning dataset is returned, pls refer to [here](https://github.com/rapidsai/cuvs/blob/branch-25.02/cpp/include/cuvs/neighbors/common.hpp#L234). So, I unintentionally did the right thing.